grace/evaluation/process.py

```diff
@@ -28,30 +30,68 @@ def update_graph_with_dummy_predictions(
     Notes
     -----
     - Modifies the graph in place.
-    - Only motifs with the same object identity will be linked.
+    - The 'uncertainty' parameter separates the distributions of predictions
+        for the respective classes, e.g. 0.5 will produce a distribution of
+        attribute values for TN between [0.0, 0.5) & TP between (0.5, 1.0]
+        which are very close, whilst 0.1 will force larger separation between
+        them, i.e. TN between [0.0, 0.1) & TP between (0.9, 1.0].
+    - Only motifs with the same object identity should be linked (TODO)
     """
 
     # Make sure all objects are labelled according to object identity:
     nodes = list(G.nodes.data())
-    obj_identity = ["object_idx" in node for _, node in nodes]
-    assert all(obj_identity)
+    obj_identity = ["object_idx" in n and "label" in n for _, n in nodes]
+    lbl_identity = [GraphAttrs.NODE_GROUND_TRUTH in n for _, n in nodes]
+    assert all(obj_identity) or all(lbl_identity)
 
     for _, node in nodes:
-        pd = np.random.random() * node_confidence
-        if node["label"] > 0:
-            node[GraphAttrs.NODE_PREDICTION] = pd
-        else:
-            node[GraphAttrs.NODE_PREDICTION] = 1 - pd
+        pd = np.random.random() * node_uncertainty
+
+        if all(obj_identity):
+            if node["label"] > 0:  # true positive node
+                node[GraphAttrs.NODE_PREDICTION] = (1, np.array([pd, 1 - pd]))
+            else:
+                node[GraphAttrs.NODE_PREDICTION] = (0, np.array([1 - pd, pd]))
+
+        if all(lbl_identity):
+            if node[GraphAttrs.NODE_GROUND_TRUTH] == Annotation.TRUE_POSITIVE:
+                node[GraphAttrs.NODE_PREDICTION] = (1, np.array([pd, 1 - pd]))
+            else:
+                node[GraphAttrs.NODE_PREDICTION] = (0, np.array([1 - pd, pd]))
 
     for edge in G.edges.data():
-        pd = np.random.random() * edge_confidence
+        pd = np.random.random() * edge_uncertainty
         _, e_i = nodes[edge[0]]
         _, e_j = nodes[edge[1]]
 
-        if e_i["object_idx"] == e_j["object_idx"] and e_i["label"] > 0:
-            edge[2][GraphAttrs.EDGE_PREDICTION] = 1 - pd
-        else:
-            edge[2][GraphAttrs.EDGE_PREDICTION] = pd
+        if all(obj_identity):
+            if (
+                e_i["object_idx"] == e_j["object_idx"] and e_i["label"] > 0
+            ):  # TP
+                edge[-1][GraphAttrs.EDGE_PREDICTION] = (
+                    1,
+                    np.array([pd, 1 - pd]),
+                )
+            else:
+                edge[-1][GraphAttrs.EDGE_PREDICTION] = (
+                    0,
+                    np.array([1 - pd, pd]),
+                )
```

Comment:
Looking at the code further down (in the optimiser), I'm unsure why you need to store the probabilities as an array?

Reply:
You don't, but that's how the label (=argmax) is calculated. It's stored for convenience and allows quick checks if the sum adds up to 1.